README.md
purple-line
===========

libpurple (Pidgin, Finch) protocol plugin for [LINE](http://line.me/).

![Screenshot](http://i.imgur.com/By1yLXB.png)

Install via package manager (Ubuntu/Debian)
-------------------------------------------

An APT repository is available for installing the plugin. The repository contains the purple-line
package itself, and the required Apache Thrift packages which are not properly packaged by either
distribution.

* http://debian.altrepo.eu/ (main)
* http://debian.surlinter.net/ (mirror)

For instructions for adding a custom repository on Ubuntu, see
[the Ubuntu wiki](https://help.ubuntu.com/community/Repositories/Ubuntu).

For Debian, see [the Debian wiki](https://www.debian.org/releases/), or just add the following line
to your `sources.list` file:

    deb http://debian.altrepo.eu/ code_name main

The code name is wriiten in
[debian.altrepo.eu wiki](http://altrepo.eu/git/debian.altrepo.eu/wikis/home#note).
In order to validate package signatures you need to add
[this public key](http://debian.altrepo.eu/altrepo_eu.pub) to the APT key list. It can be done
using the following command:

    wget -qO - http://debian.altrepo.eu/altrepo_eu.pub | sudo apt-key add -

On either distribution, after adding the repository and key, run the following commands to install
the plugin:

    sudo apt-get update
    sudo apt-get install purple-line

Install from source (Ubuntu/Debian)
-----------------------------------

apt-build enables you to install from source easily.
You need to add source entry to /etc/apt/sources.list.

    sudo apt-get install apt-build
    echo 'deb-src http://debian.altrepo.eu/ code_name main' >> /etc/apt/sources.list
    sudo apt-build install purple-line
    

Note that apt-build is not official apt family.
If you install via source, your purple-line version is
up to date with git repository.So, make sure that the installation via source is not tested.
Compile will fail if build dependencies are not correctly set.

Install from source (Arch Linux)
--------------------------------

Arch Linux packages all the required dependencies and there's a PKGBUILD available (not yet in AUR).
You can install the plugin by simply typing:

    sudo pacman -S base-devel
    curl -O http://altrepo.eu/git/arch.altrepo.eu/raw/master/purple-line/PKGBUILD
    makepkg -s -i -c

Install from source (any distribution)
--------------------------------------

Make sure you have the required prerequisites installed:

* libpurple - Library that provides the core functionality of Pidgin and other compatible clients.
  Probably available from your package manager.
* thrift - Apache Thrift compiler. May be available from your package manager.
* libthrift - Apache Thrift C++ library. May be available from your package manager.
* libgcrypt - Crypto library. Probably available from your package manager.

To install the plugin system-wide, run:

    make
    sudo make install

The makefile supports a flag THRIFT_STATIC=true which causes it to download and build a version of
Thrift and statically link it. This should be convenient for people using one of the numerous
distributions that do not package Thrift.

You can also install the plugin for your user only by replacing `install` with `user-install`.

Features implemented
--------------------

* Logging in
  * Authentication
  * Fetching user profile
  * Account icon
  * Syncing friends, groups and chats
* Send and receive messages in IM, groups and chats
* Fetch recent messages
  * For groups and chats
  * For IMs
* Synchronize buddy list on the fly
  * Adding friends
  * Blocking friends
  * Removing friends
  * Joining chats
  * Leaving chats
  * Group invitations
  * Joining groups
  * Leaving groups
* Buddy icons
* Editing buddy list
 * Removing friends
 * Leaving chats
 * Leaving groups
* Message types
 * Text (send/receive)
 * Sticker (send via command/receive)
 * Image (send/receive)
 * Audio (send/receive)
 * Location (receive)

Features not yet implemented
----------------------------

* Only fetch unseen messages, let a log plugin handle already seen messages
* Implement timeouts for faster reconnections
* Synchronize buddy list on the fly
  * Sync group/chat users more gracefully, show people joining/leaving
* Editing buddy list
  * Adding friends (needs search function)
  * Creating chats
  * Inviting people to chats
  * Creating groups
  * Updating groups
  * Inviting people to groups
* Changing your account icon
* Message types
  * Audio/Video (send) File transfer API for sending?
  * Figure out what the other 15 message types mean...
* Emoji (is it possible to tap into the smiley system for sending too?)
* Companion Pidgin plugin
  * "Show more history" button
  * Sticker list
  * Open image messages
  * Open audio messages
  * Open video messages
* Sending/receiving "message read" notifications
* Check builds on more platforms
* Packaging


debian/changelog
purple-line (0.1.2-ubuntu2) vivid; urgency=low

    * Internal LINE version updated.Fix new thrift bug.

 -- Akira Nakagawa <matyapiro31@gmail.com>  Wed, 28 Oct 2015 20:30:00 +0900
purple-line (0.1.1-ubuntu2) vivid; urgency=low

    * Support new LINE login.

 -- Akira Nakagawa <matyapiro31@gmail.com>  Thu, 17 Sep 2015 21:04:00 +0900
purple-line (0.1-ubuntu1) vivid; urgency=low

    * Release for public.

 -- Akira Nakagawa <matyapiro31@gmail.com>  Tue, 28 Jul 2015 13:42:00 +0900
purple-line (0.0-ubuntu1) trusty; urgency=low

    * Initial Release

 -- Akira Nakagawa <matyapiro31@gmail.com>  Sat, 25 Jul 2015 21:31:57 +0900



libpurple/poller.cpp
#include <time.h>

#include <debug.h>

#include "constants.hpp"
#include "poller.hpp"
#include "purpleline.hpp"

Poller::Poller(PurpleLine &parent)
    : parent(parent)
{
    client = boost::make_shared<ThriftClient>(parent.acct, parent.conn, LINE_POLL_PATH);
    client->set_auto_reconnect(true);
}

Poller::~Poller() {
    client.reset();
}

void Poller::start() {
    fetch_operations();
}

void Poller::fetch_operations() {
    client->send_fetchOperations(local_rev, 50);
    client->send([this]() {
        int status = client->status_code();

        if (status == -1) {
            // Plugin closing
            return;
        } else if (status == 410) {
            // Long poll timeout, resend
            fetch_operations();
            return;
        } else if (status != 200) {
            purple_debug_warning("line", "fetchOperations error %d. TODO: Retry after a timeout", status);
            return;
        }

        std::vector<line::Operation> operations;
        client->recv_fetchOperations(operations);

        for (line::Operation &op: operations) {
            switch (op.type) {
                case line::OpType::END_OF_OPERATION: // 0
                    break;

                case line::OpType::ADD_CONTACT: // 4
                    parent.blist_update_buddy(op.param1);
                    break;

                case line::OpType::BLOCK_CONTACT: // 6
                    parent.blist_remove_buddy(op.param1);
                    break;

                case line::OpType::UNBLOCK_CONTACT: // 7
                    parent.blist_update_buddy(op.param1);
                    break;

                case line::OpType::CREATE_GROUP: // 9
                case line::OpType::UPDATE_GROUP: // 10
                case line::OpType::NOTIFIED_UPDATE_GROUP: // 11
                case line::OpType::INVITE_INTO_GROUP: // 12
                    parent.blist_update_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::NOTIFIED_INVITE_INTO_GROUP: // 13
                    op_notified_invite_into_group(op);
                    break;

                case line::OpType::LEAVE_GROUP: // 14
                    parent.blist_remove_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::NOTIFIED_LEAVE_GROUP: // 15
                    parent.blist_update_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::ACCEPT_GROUP_INVITATION: // 16
                    parent.blist_update_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::NOTIFIED_ACCEPT_GROUP_INVITATION: // 17
                case line::OpType::KICKOUT_FROM_GROUP: // 18
                    parent.blist_update_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::NOTIFIED_KICKOUT_FROM_GROUP: // 19
                    op_notified_kickout_from_group(op);
                    break;

                case line::OpType::CREATE_ROOM: // 20
                case line::OpType::INVITE_INTO_ROOM: // 21
                    parent.blist_update_chat(op.param1, ChatType::ROOM);
                    break;

                case line::OpType::NOTIFIED_INVITE_INTO_ROOM: // 22
                    // TODO: Perhaps show who invited the user (param2)
                    parent.blist_update_chat(op.param1, ChatType::ROOM);
                    break;

                case line::OpType::LEAVE_ROOM: // 23
                    parent.blist_remove_chat(op.param1, ChatType::ROOM);
                    break;

                case line::OpType::NOTIFIED_LEAVE_ROOM: // 24
                    parent.blist_update_chat(op.param1, ChatType::ROOM);

                case line::OpType::SEND_MESSAGE: // 25
                case line::OpType::RECEIVE_MESSAGE: // 26
                    parent.write_message(op.message, false);
                    break;

                case line::OpType::CANCEL_INVITATION_GROUP: // 31
                case line::OpType::NOTIFIED_CANCEL_INVITATION_GROUP: // 32
                    parent.blist_update_chat(op.param1, ChatType::GROUP);
                    break;

                case line::OpType::DUMMY: // 48;
                    break;

                case line::OpType::UPDATE_CONTACT: // 49
                    parent.blist_update_buddy(op.param1);
                    break;

                default:
                    purple_debug_warning("line", "Unhandled operation type: %d\n", op.type);
                    break;
            }

            if (op.revision > local_rev)
                local_rev = op.revision;
        }

        fetch_operations();
    });
}

void Poller::op_notified_kickout_from_group(line::Operation &op) {
    std::string msg;

    if (op.param3 == parent.profile.mid) {
        msg = "You were removed from the group by ";
        parent.blist_remove_chat(op.param1, ChatType::GROUP);
    } else {
        msg = "Removed from the group by ";
        parent.blist_update_chat(op.param1, ChatType::GROUP);
    }

    if (parent.contacts.count(op.param2) == 1)
        msg += parent.contacts[op.param2].displayName;
    else
        msg += "(unknown contact)";

    PurpleConversation *conv = purple_find_conversation_with_account(
        PURPLE_CONV_TYPE_CHAT,
        op.param1.c_str(),
        parent.acct);

    if (conv) {
        purple_conversation_write(
            conv,
            op.param3.c_str(),
            msg.c_str(),
            PURPLE_MESSAGE_SYSTEM,
            time(NULL));
    }
}

void Poller::op_notified_invite_into_group(line::Operation &op) {
    // TODO: Maybe use cached objects instead of re-requesting every time

    parent.c_out->send_getGroup(op.param1);
    parent.c_out->send([this, op]{
        line::Group group;
        parent.c_out->recv_getGroup(group);

        if (!group.__isset.id) {
            purple_debug_warning("line", "Invited into unknown group: %s\n", op.param1.c_str());
            return;
        }

        parent.c_out->send_getContact(op.param2);
        parent.c_out->send([this, group, op]() mutable {
            line::Contact inviter;
            parent.c_out->recv_getContact(inviter);

            parent.c_out->send_getContact(op.param3);
            parent.c_out->send([this, group, inviter, op]() mutable {
                line::Contact invitee;
                parent.c_out->recv_getContact(invitee);

                parent.handle_group_invite(group, invitee, inviter);
            });
        });
    });
}
